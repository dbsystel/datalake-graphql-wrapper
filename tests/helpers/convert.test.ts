import { expect, test, describe } from 'vitest'
import {
  parseFields,
  generateJsonSchema,
  generateModel,
  convertField,
} from '../../src/helpers/convert'
import {
  generatedJsonSchema,
  generatedTableColumns,
  parsedFields,
} from '../mockdata'

describe('Helpers - Convert', () => {
  test('parseFields', async () => {
    const fnResult = parseFields(generatedTableColumns)
    expect(fnResult).toMatchSnapshot()
  })

  test('generateJsonSchema', async () => {
    const fnResult = generateJsonSchema({ source: parsedFields })
    expect(fnResult).toMatchSnapshot()
  })

  test('generateModel', async () => {
    const fnResult = generateModel({
      source: generatedJsonSchema,
      name: 'mockModel',
      models: {},
      isRoot: true,
    })

    expect(fnResult).toMatchSnapshot()
    expect(Object.keys(fnResult).length).toBe(4)
    expect(Object.keys(fnResult).sort()).toEqual([
      'mockModel',
      'mockModel_Complexobject',
      'mockModel_Cosub3',
      'mockModel_Simpleobject',
    ])
  })
})

describe('Helpers - Convert - Field Converter', () => {
  test('stringField', () => {
    const fnResult = convertField({
      fieldName: 'stringfield',
      fieldDefinition: generatedJsonSchema.properties.stringfield,
    })
    expect(fnResult).toMatchObject({
      graphqlType: 'String',
      graphqlTplType: "'String'",
      fieldType: 'string',
      subModel: false,
      filter: true,
    })
  })
  test('numberField', () => {
    const fnResult = convertField({
      fieldName: 'numberfield',
      fieldDefinition: generatedJsonSchema.properties.bigintfield,
    })
    expect(fnResult).toMatchObject({
      graphqlType: 'Int',
      graphqlTplType: "'Int'",
      fieldType: 'number',
      subModel: false,
      filter: true,
    })
  })
  test('booleanField', () => {
    const fnResult = convertField({
      fieldName: 'booleanfield',
      fieldDefinition: generatedJsonSchema.properties.booleanfield,
    })
    expect(fnResult).toMatchObject({
      graphqlType: 'Boolean',
      graphqlTplType: "'Boolean'",
      fieldType: 'boolean',
      subModel: false,
      filter: true,
    })
  })
  test('datetimeField', () => {
    const fnResult = convertField({
      fieldName: 'datetimefield',
      fieldDefinition: generatedJsonSchema.properties.datetimefield,
    })
    expect(fnResult).toMatchObject({
      graphqlType: 'DateTime',
      graphqlTplType: "'DateTime'",
      fieldType: 'Date',
      subModel: false,
      filter: true,
    })
  })
  test('arrayStringField', () => {
    const fnResult = convertField({
      fieldName: 'arraystringfield',
      fieldDefinition: generatedJsonSchema.properties.stringarray,
    })
    expect(fnResult).toMatchObject({
      graphqlType: '[String]',
      graphqlTplType: "['String']",
      fieldType: 'string[]',
      subModel: false,
      isArray: true,
      filter: false,
    })
  })
  test('objectField', () => {
    const fnResult = convertField({
      fieldName: 'objectfield',
      fieldDefinition: generatedJsonSchema.properties.simpleobject,
      parent: 'mockModel',
    })
    expect(fnResult).toMatchObject({
      graphqlTplType: 'mockModel_Objectfield',
      graphqlType: 'mockModel_Objectfield',
      interfaceName: 'mockModel_ObjectfieldInterface',
      subModel: {
        additionalProperties: false,
        properties: {
          sosub1: {
            type: 'string',
          },
          sosub2: {
            type: 'integer',
          },
        },
        type: 'object',
      },
    })
  })
})
