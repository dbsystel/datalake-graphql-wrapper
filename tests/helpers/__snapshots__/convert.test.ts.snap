// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Helpers - Convert > generateJsonSchema 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "bigintfield": {
      "type": "integer",
    },
    "booleanfield": {
      "type": "boolean",
    },
    "complexobject": {
      "additionalProperties": false,
      "properties": {
        "cosub1": {
          "type": "string",
        },
        "cosub2": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "cosub3": {
          "items": {
            "additionalProperties": false,
            "properties": {
              "cosub31": {
                "type": "string",
              },
              "cosub32": {
                "type": "string",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
    "dash-field": {
      "type": "string",
    },
    "datetimefield": {
      "type": "date-time",
    },
    "simpleobject": {
      "additionalProperties": false,
      "properties": {
        "sosub1": {
          "type": "string",
        },
        "sosub2": {
          "type": "integer",
        },
      },
      "type": "object",
    },
    "special-aou": {
      "type": "string",
    },
    "stringarray": {
      "items": {
        "type": "string",
      },
      "type": "array",
    },
    "stringfield": {
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`Helpers - Convert > generateModel 1`] = `
{
  "mockModel": {
    "fields": {
      "bigintfield": {
        "fieldType": "number",
        "filter": true,
        "graphqlTplType": "'Int'",
        "graphqlType": "Int",
        "interfaceName": undefined,
        "isArray": false,
        "name": "bigintfield",
        "nullable": true,
      },
      "booleanfield": {
        "fieldType": "boolean",
        "filter": true,
        "graphqlTplType": "'Boolean'",
        "graphqlType": "Boolean",
        "interfaceName": undefined,
        "isArray": false,
        "name": "booleanfield",
        "nullable": true,
      },
      "complexobject": {
        "fieldType": "mockModel_Complexobject",
        "filter": false,
        "graphqlTplType": "mockModel_Complexobject",
        "graphqlType": "mockModel_Complexobject",
        "interfaceName": "mockModel_ComplexobjectInterface",
        "isArray": false,
        "name": "complexobject",
        "nullable": true,
      },
      "dash-field": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "dash-field",
        "nullable": true,
      },
      "datetimefield": {
        "fieldType": "Date",
        "filter": true,
        "graphqlTplType": "'DateTime'",
        "graphqlType": "DateTime",
        "interfaceName": undefined,
        "isArray": false,
        "name": "datetimefield",
        "nullable": true,
      },
      "simpleobject": {
        "fieldType": "mockModel_Simpleobject",
        "filter": false,
        "graphqlTplType": "mockModel_Simpleobject",
        "graphqlType": "mockModel_Simpleobject",
        "interfaceName": "mockModel_SimpleobjectInterface",
        "isArray": false,
        "name": "simpleobject",
        "nullable": true,
      },
      "special-aou": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "special-aou",
        "nullable": true,
      },
      "stringarray": {
        "fieldType": "string[]",
        "filter": false,
        "graphqlTplType": "['String']",
        "graphqlType": "[String]",
        "interfaceName": undefined,
        "isArray": true,
        "name": "stringarray",
        "nullable": true,
      },
      "stringfield": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "stringfield",
        "nullable": true,
      },
    },
    "interfaceName": "mockModelInterface",
    "modelName": "mockModel",
    "root": true,
  },
  "mockModel_Complexobject": {
    "fields": {
      "cosub1": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "cosub1",
        "nullable": true,
      },
      "cosub2": {
        "fieldType": "string[]",
        "filter": false,
        "graphqlTplType": "['String']",
        "graphqlType": "[String]",
        "interfaceName": undefined,
        "isArray": true,
        "name": "cosub2",
        "nullable": true,
      },
      "cosub3": {
        "fieldType": "mockModel_Cosub3[]",
        "filter": false,
        "graphqlTplType": "[mockModel_Cosub3]",
        "graphqlType": "[mockModel_Cosub3]",
        "interfaceName": "mockModel_Cosub3Interface[]",
        "isArray": true,
        "name": "cosub3",
        "nullable": true,
      },
    },
    "interfaceName": "mockModel_ComplexobjectInterface",
    "modelName": "mockModel_Complexobject",
    "root": false,
  },
  "mockModel_Cosub3": {
    "fields": {
      "cosub31": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "cosub31",
        "nullable": true,
      },
      "cosub32": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "cosub32",
        "nullable": true,
      },
    },
    "interfaceName": "mockModel_Cosub3Interface",
    "modelName": "mockModel_Cosub3",
    "root": false,
  },
  "mockModel_Simpleobject": {
    "fields": {
      "sosub1": {
        "fieldType": "string",
        "filter": true,
        "graphqlTplType": "'String'",
        "graphqlType": "String",
        "interfaceName": undefined,
        "isArray": false,
        "name": "sosub1",
        "nullable": true,
      },
      "sosub2": {
        "fieldType": "number",
        "filter": true,
        "graphqlTplType": "'Int'",
        "graphqlType": "Int",
        "interfaceName": undefined,
        "isArray": false,
        "name": "sosub2",
        "nullable": true,
      },
    },
    "interfaceName": "mockModel_SimpleobjectInterface",
    "modelName": "mockModel_Simpleobject",
    "root": false,
  },
}
`;

exports[`Helpers - Convert > parseFields 1`] = `
{
  "bigintfield": "bigint",
  "booleanfield": "boolean",
  "complexobject": {
    "cosub1": "varchar",
    "cosub2": [
      "varchar",
    ],
    "cosub3": [
      {
        "cosub31": "varchar",
        "cosub32": "varchar",
      },
    ],
  },
  "dash-field": "varchar",
  "datetimefield": "timestamp(3)",
  "simpleobject": {
    "sosub1": "varchar",
    "sosub2": "bigint",
  },
  "special-aou": "varchar",
  "stringarray": [
    "varchar",
  ],
  "stringfield": "varchar",
}
`;
